File: dp_labs_and_solutions/original/dp_labs_and_solutions/dp_matlab_labs_and_solutions/chain_of_responsibility/solution/Handler.m
% Handler.m

classdef (Abstract) Handler < handle
	properties (GetAccess = private, SetAccess = private)
		Successor % Handler
	end

	methods
		% constructor:
		function theHandler = Handler(aSuccessor)
			theHandler.Successor = aSuccessor;
		end
	end

	methods (Abstract, Access = protected)
		result = canHandle(theHandler, aCustomerRequestsDiscountEvent)

		doHandle(theHandler, aCustomerRequestsDiscountEvent)
	end

	methods (Sealed)
		function handle(theHandler, aCustomerRequestsDiscountEvent)
			if theHandler.canHandle(aCustomerRequestsDiscountEvent)
				theHandler.doHandle(aCustomerRequestsDiscountEvent);
			elseif ~isempty(theHandler.Successor)
				theHandler.Successor.handle(aCustomerRequestsDiscountEvent);
			else
				% the event is not handled.
			end
		end
	end
end

File: dp_labs_and_solutions/original/dp_labs_and_solutions/dp_java_labs_and_solutions/chain_of_responsibility/solution/Handler.java
// Handler.java

public abstract class Handler
{
	public Handler(Handler aSuccessor)
	{
		setSuccessor(aSuccessor);
	}

	public final void handle(CustomerRequestsDiscountEvent aCustomerRequestsDiscountEvent)
	{
		if (canHandle(aCustomerRequestsDiscountEvent))
		{
			doHandle(aCustomerRequestsDiscountEvent);
		}
		else if (getSuccessor() != null)
		{
			getSuccessor().handle(aCustomerRequestsDiscountEvent);
		}
		else
		{
			// aCustomerRequestsDiscountEvent is not handled.
		}
	}

	protected abstract boolean canHandle(CustomerRequestsDiscountEvent aCustomerRequestsDiscountEvent);

	protected abstract void doHandle(CustomerRequestsDiscountEvent aCustomerRequestsDiscountEvent);

	private Handler getSuccessor()
	{
		return mySuccessor;
	}

	private void setSuccessor(Handler aSuccessor)
	{
		mySuccessor = aSuccessor;
	}

	private Handler mySuccessor;
}
